lua/neo-tree/utils/init.lua

```diff
@@ -214,39 +218,78 @@ M.human_size = function (size)
   return human
 end
 
----Gets diagnostic severity counts for all files
----@return table table { file_path = { Error = int, Warning = int, Information = int, Hint = int, Unknown = int } }
+---Gets non-zero diagnostics counts for each open file and each ancestor directory.
+---severity_number and severity_string refer to the highest severity with
+---non-zero diagnostics count.
+---Entry is nil if all counts are 0
+---@return table table
+---{ [file_path] = {
+---    severity_number = int,
+---    severity_string = string,
+---    Error = int or nil,
+---    Warn = int or nil,
+---    Info = int or nil
+---    Hint = int or nil,
+---  } or nil }
 M.get_diagnostic_counts = function()
-  local d = vim.diagnostic.get()
   local lookup = {}
-  for _, diag in ipairs(d) do
-    if diag.source == "Lua Diagnostics." and diag.message == "Undefined global `vim`." then
-      -- ignore this diagnostic
-    else
-      local success, file_name = pcall(vim.api.nvim_buf_get_name, diag.bufnr)
-      if success then
-        local sev = diag_severity_to_string(diag.severity)
-        if sev then
-          local entry = lookup[file_name] or { severity_number = 4 }
-          entry[sev] = (entry[sev] or 0) + 1
-          entry.severity_number = math.min(entry.severity_number, diag.severity)
-          entry.severity_string = diag_severity_to_string(entry.severity_number)
-          lookup[file_name] = entry
+
+  for ns, _ in pairs(vim.diagnostic.get_namespaces()) do
+    for _, bufnr in ipairs(vim.api.nvim_list_bufs()) do
+      local success, file_name = pcall(vim.api.nvim_buf_get_name, bufnr)
+      if success and not vim.diagnostic.is_disabled(bufnr, ns) then
+        for severity, _ in ipairs(vim.diagnostic.severity) do
+          local diagnostics = vim.diagnostic.get(bufnr, { namespace = ns, severity = severity })
+
+          -- It would be nice to not do this manual filtering below
+          -- Note that lua diagnostics can be suppressed using neodev
+          -- or lua_ls settings instead.
+          -- https://github.com/folke/neodev.nvim
```

Comment:
I totally agree with this **and** I think it's time to just remove it. This was added before neodev and before I learned about the lua_ls setting.

It's a minor breaking change, but it has such an obvious solution and will remove a lot of wasted work so I think it is worth it.